src/raftstore/store/peer.rs

```diff
@@ -955,40 +938,50 @@ impl Peer {
         }
     }
 
+    // All updates about leader_lease must be done by this method.
+    // And only leader can call it.
+    fn step_leader_lease(&mut self, expired_time: Timespec, safe: bool) {
+        assert!(
+            self.is_leader(),
+            "{} non-leader peer {} try to update leader lease",
+            self.tag,
+            self.peer.get_id()
+        );
+        if safe {
+            self.leader_lease.update_upper(expired_time);
+        } else {
+            self.leader_lease.update_lower(expired_time);
+        }
+    }
+
+    fn expire_leader_lease(&mut self) {
+        self.leader_lease.clear();
+    }
+
     fn update_lease_with(&mut self, propose_time: Timespec) {
+        // A nonleader peer should never has leader lease.
+        if !self.is_leader() {
+            return;
+        }
         // Try to renew the leader lease as this command asks to.
-        if self.leader_lease_expired_time.is_some() {
-            let current_expired_time =
-                match self.leader_lease_expired_time.as_ref().unwrap().as_ref() {
-                    Either::Left(safe_expired_time) => *safe_expired_time,
-                    Either::Right(unsafe_expired_time) => *unsafe_expired_time,
-                };
-            // This peer is leader and has recorded leader lease.
-            // Calculate the renewed lease for this command. If the renewed lease lives longer
-            // than the current leader lease, update the current leader lease to the renewed lease.
-            let next_expired_time = self.next_lease_expired_time(propose_time);
-            // Use the lease expired timestamp comparison here, so that these codes still
-            // work no matter how the leader changes before applying this command.
-            if current_expired_time < next_expired_time {
+        let next_expired_time = self.next_lease_expired_time(propose_time);
+        // If the renewed lease lives longer than the current leader lease,
+        // update the current leader lease to the renewed lease.
+        let behind_lower = self.leader_lease.behind_lower(next_expired_time);
```

Comment:
The original implementation compare it with the `next_expired_time`, but I think it should be `monotonic_raw_now()`. Otherwise, the lease may become safe while it is still behind the unsafe lowerbound.
cc @BusyJay @siddontang